dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,239 @@
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `FailedRequestClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = FailedRequestClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
+    this.captureFailedRequests = true,
+    Client? client,
+    Hub? hub,
+  })  : _hub = hub ?? HubAdapter(),
+        _client = client ?? Client();
+
+  final Client _client;
+  final Hub _hub;
+
+  /// Configures wether to record exceptions for failed requests.
+  /// Examples for captures exceptions are:
+  /// - In an browser environment this can be requests which fail because of CORS.
+  /// - In an mobile or desktop application this can be requests which failed
+  ///   because the connection was interrupted.
+  final bool captureFailedRequests;
+
+  /// Configures up to which size request bodies should be included in events.
+  /// This does not change wether an event is captured.
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<SentryStatusCode> failedRequestStatusCodes;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    int? statusCode;
+    Object? exception;
+    StackTrace? stackTrace;
+
+    final stopwatch = Stopwatch();
+    stopwatch.start();
+
+    try {
+      final response = await _client.send(request);
+      statusCode = response.statusCode;
+      return response;
+    } catch (e, st) {
+      exception = e;
+      stackTrace = st;
+      rethrow;
+    } finally {
+      stopwatch.stop();
+
+      if (captureFailedRequests) {
+        await _captureEvent(
+          exception: exception,
+          stackTrace: stackTrace,
+          request: request,
+          requestDuration: stopwatch.elapsed,
+        );
+      }
+
+      if (failedRequestStatusCodes.containsStatusCode(statusCode)) {
+        // Capture an exception if the status code is considered bad
+        await _captureEvent(
+          request: request,
+          reason: failedRequestStatusCodes.toDescription(),
+          requestDuration: stopwatch.elapsed,
+        );
+      }
```

Comment:
does it mean its gonna capture 2 events? I see 2 calls to `_captureEvent`

Reply:
should it be either if there was an exception or if it was within the given range?